File: server/src/main/java/gg/mineral/server/network/packet/play/clientbound/ParticlePacket.java
package gg.mineral.server.network.packet.play.clientbound;

import gg.mineral.api.network.packet.Packet;

import io.netty.buffer.ByteBuf;

public final record ParticlePacket(String particleName, float x, float y, float z, float offsetX, float offsetY,
        float offsetZ, float particleData, int numberOfParticles) implements Packet.OUTGOING {
    @Override
    public void serialize(ByteBuf os) {
        writeString(os, particleName);
        os.writeFloat(x);
        os.writeFloat(y);
        os.writeFloat(z);
        os.writeFloat(offsetX);
        os.writeFloat(offsetY);
        os.writeFloat(offsetZ);
        os.writeFloat(particleData);
        os.writeInt(numberOfParticles);
    }

    @Override
    public byte getId() {
        return 0x2A;
    }
}

File: server/src/main/kotlin/gg/mineral/server/network/packet/play/clientbound/ParticlePacket.kt
package gg.mineral.server.network.packet.play.clientbound

import gg.mineral.api.network.packet.Packet
import io.netty.buffer.ByteBuf

@JvmRecord
data class ParticlePacket(
    val particleName: String, val x: Float, val y: Float, val z: Float, val offsetX: Float, val offsetY: Float,
    val offsetZ: Float, val particleData: Float, val numberOfParticles: Int
) : Packet.OUTGOING {
    override fun serialize(os: ByteBuf) {
        os.writeString(particleName)
        os.writeFloat(x, y, z, offsetX, offsetY, offsetZ, particleData)
        os.writeInt(numberOfParticles)
    }

    override val id: Byte
        get() = 0x2A
}
